### ISSUE TITLE
Wrong error triggered in calls with keyword arguments

### ISSUE THREAD

**Issue by pschella** [pybind/pybind11#688]
When a class with an overloaded copy constructor:

```
struct Bla {
    Bla() = default;
    explicit Bla(const Bla &, bool deep=true) { }
    explicit Bla(const Bla &, int i, std::string s="", bool deep=true) { }
};
```

wrapped as

```
    py::class_<Bla>(m, "Bla")
        .def(py::init<>())
        .def(py::init<const Bla &, bool>(),
                "other"_a, "deep"_a=true)
        .def(py::init<const Bla &, int, std::string, bool>(),
                "other"_a, "i"_a, "s"_a="", "deep"_a=true);
```

Is called with:

```
b = Bla()
c = Bla(b, 1, deep=False)
```

The error message reads: `TypeError: __init__(): got multiple values for argument 'deep'`. Which is confusing.

**Comment by pschella** [comment]
Note that an error is expected (because it is called with the wrong number of arguments), but this particular error is surprising.

**Comment by pschella** [comment]
It is actually worse than I thought. Not only is the error confusing. It actually shouldn't error at all when the missing argument has a default value. But it does...

**Comment by jagerman** [comment]
Hmm, might be related to #643 and/or #634.  I'll take a look.

**Comment by jagerman** [comment]
(Is this happening with 2.0.1?)

**Comment by pschella** [comment]
It works on 2.0.1.

**Comment by pschella** [comment]
I.e. is broken on master but not on 2.0.1.

**Comment by jagerman** [comment]
Bisected to a3a020c1b3e99f2cde28e9d03b4aa9951b2e71e6 (PR #611).

**Comment by jagerman** [comment]
I should have known that something I assumed to be "highly likely" would fail before long (see the diff).  Fixed in #696.
